torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    max_entries_ = parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0);
+    enabled_ = max_entries_ > 0;
+  }
+  using EventList = std::vector<at::cuda::CUDAEvent>;
+  struct Entry {
+    size_t id_;
+    size_t pg_id_;
+    size_t seq_id_;
+    const char* profiling_name_;
+    c10::SmallVector<int64_t, 4> sizes_;
+    std::shared_ptr<torch::CapturedTraceback> traceback_;
+    EventList *start_, *end_;
+    const char* state_ = "scheduled";
+  };
+
+  bool enabled_ = false;
+  std::mutex mutex_;
+  std::vector<Entry> entries_;
+  size_t max_entries_ = 1024 * 16;
+  size_t next_ = 0;
+  size_t id_ = 0;
+
+  c10::optional<size_t> record(
+      size_t pg_id,
+      size_t seq_id,
+      const char* profiling_name,
+      const c10::optional<std::vector<at::Tensor>>& inputs,
+      EventList* start,
+      EventList* end) {
+    if (!enabled_) {
+      return c10::nullopt;
+    }
+    auto traceback = torch::CapturedTraceback::gather(true, true, true);
+    std::lock_guard<std::mutex> guard(mutex_);
+    at::IntArrayRef sizes;
+    if (inputs && inputs->size() > 0) {
+      sizes = inputs->at(0).sizes();
+    }
+    auto te = Entry{
+        id_,
+        pg_id,
+        seq_id,
+        profiling_name,
+        c10::SmallVector<int64_t, 4>(sizes.begin(), sizes.end()),
+        std::move(traceback),
+        std::move(start),
+        std::move(end)};
+    if (entries_.size() < max_entries_) {
+      entries_.emplace_back(std::move(te));
+    } else {
+      entries_[next_++] = std::move(te);
+      if (next_ == max_entries_) {
+        next_ = 0;
+      }
+    }
+    return id_++;
+  }
+
+  std::vector<Entry> report() {
+    std::lock_guard<std::mutex> guard(mutex_);
+    std::vector<Entry> result;
+    result.reserve(entries_.size());
+    result.insert(result.end(), entries_.begin() + next_, entries_.end());
+    result.insert(result.end(), entries_.begin(), entries_.begin() + next_);
+    // query any remaining events
+    for (auto& r : result) {
+      if (r.start_ != nullptr) {
+        bool started = true;
+        for (auto& ev : *r.start_) {
+          if (!ev.query()) {
+            started = false;
+            break;
+          }
+        }
+        if (started) {
+          r.state_ = "started";
+        }
+        r.start_ = nullptr;
+      }
+      if (r.end_ != nullptr) {
+        bool completed = true;
+        for (auto& ev : *r.end_) {
+          if (!ev.query()) {
+            completed = false;
+            break;
+          }
+        }
+        if (completed) {
+          r.state_ = "completed";
+        }
+        r.end_ = nullptr;
+      }
+    }
+    return result;
+  }
+
+  void complete(c10::optional<size_t> id) {
+    if (!enabled_ || !id) {
+      return;
+    }
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& entry = entries_.at(*id % max_entries_);
+    if (entry.id_ == *id) {
+      entry.state_ = "completed";
+      entry.start_ = entry.end_ = nullptr;
+    }
+  }
+
+  std::string dump() {
```

Comment:
bikeshedding: dump and report both kind of sound like synonyms to me.  report actually doesn't 'generate a report' it just updates the state of the state of the events inside the buffer to the latest info it can get.  Maybe leave dump as is but change report to something more descriptive?

Reply:
I named report `dump_entries`. It really only exists to have a function that lives inside of the lock, and another that lives outside of it. `symbolize` is expensive, so it needs to be outside that lock.